crates/telio-nurse/src/aggregator.rs

```diff
@@ -0,0 +1,113 @@
+use std::{
+    collections::{btree_map::Entry, BTreeMap},
+    time::Instant,
+};
+
+use parking_lot::RwLock;
+use telio_crypto::PublicKey;
+use telio_model::api_config::EndpointProvider;
+use telio_utils::telio_log_warn;
+
+// Possible connection types - relayed ones allow adding info about the other endpoint
+#[derive(Copy, Clone, Debug)]
+pub enum ConnectionType {
+    Relayed,
+    Local(Option<EndpointProvider>),
+    UPnP(Option<EndpointProvider>),
+    UDPHolePunch(Option<EndpointProvider>),
+}
+
+// Possible disconnection reasons
+#[derive(Copy, Clone, Debug)]
+pub enum DisconnectionReason {
+    Initial,
+    DisabledByUser,
+    NetworkError,
```

Comment:
Maybe it would be nice to split `NetworkError` at least in a few subtypes:
* Forced connection termination by the server
* Timeouts
* Some OS errors (like no route to host or similar)

Reply:
Another option would be to simply include `errno` value :thinking:

Reply:
I guess it would be nicer to have it as an enum here? I mean, having an errno value forces us to maintain another mapping